spyder_kernels/customize/spydercustomize.py

```diff
@@ -917,46 +953,88 @@ def runcell(cellname, filename):
         # AttributeError --> systematically raised in Python 3
         pass
     ipython_shell = get_ipython()
-    namespace = _get_globals()
-    namespace['__file__'] = filename
     try:
-        cell_code = ipython_shell.cell_code
-    except AttributeError:
-        _print("--Run Cell Error--\n"
-               "Please use only through Spyder's Editor; "
-               "shouldn't be called manually from the console")
+        # Get code from spyder
+        cell_code = _frontend_request().run_cell(cellname, filename)
+        if not filename:
+            filename = _frontend_request().current_filename()
+    except (CommError, TimeoutError):
+        if hasattr(ipython_shell, "cell_code"):
+            # This is the old way of transmitting informations
+            cell_code = ipython_shell.cell_code
+            del ipython_shell.cell_code
+        else:
+            _print("--Run Cell Error--\n"
+                   "Please use only with Spyder; ")
+            return
+
+    if not cell_code:
+        # Nothing to execute
         return
 
+    namespace = _get_globals()
+    # Save previous __file__
+    namespace_has_file = '__file__' in namespace
+    if namespace_has_file:
+        namespace_filename = namespace['__file__']
+    namespace['__file__'] = filename
+
     # Trigger `post_execute` to exit the additional pre-execution.
     # See Spyder PR #7310.
     ipython_shell.events.trigger('post_execute')
 
     ipython_shell.run_cell(cell_code)
-    namespace.pop('__file__')
-    del ipython_shell.cell_code
+
+    # Restore __file__
+    if namespace_has_file:
+        namespace['__file__'] = namespace_filename
+    else:
+        namespace.pop('__file__')
 
 
 builtins.runcell = runcell
 
 
-def debugfile(filename, args=None, wdir=None, post_mortem=False):
+def cellcount(filename=None):
```

Comment:
Let's rename this to `cell_count`.

Reply:
Also, what happens when you pass None as `filename` here?

Reply:
From shell.get_editor: 
```
if not filename:
            return editorstack.get_current_editor()
```